enhancements/insights/container-scanner.md

```diff
@@ -0,0 +1,498 @@
+    * Value is a hash of the detected runtime version: `2.10.3.Final, 7.4.7.GA, 2.7.1, 3.6.0.redhat-00005,...` (list not exhaustive)
+
+It is not planned to add more fields to the `workloadRuntimeInfoContainer` type. However, the container scanner can be enhanced to detect more runtimes without impacting the integration in the Insights Operator.
+
+##### Examples
+
+```json
+"runtimeInfo": {
+  "os": "yqx1WBefoIAq",
+  "osVersion": "AQN-qh_OyRUA",
+  "kind": "wbpgzhNYZQOi",
+  "kindVersion": "cAsoUDy8JFNV",
+  "kindImplementer": "DzP5-GS6tDOQ",
+  "name": "zBVuVhrrC_vC",
+  "version": "lyAE7Oh63WPU"
+}
+```
+
+As the values are derived by the container scanner or correspond to software versions, it is possible to create a dictionary of the hashes and retrieve the unobfuscated values. The example above corresponds to:
+
+```json
+"runtimeInfo": {
+  "os": "rhel",
+  "osVersion": "8.9",
+  "kind": "Java",
+  "kindVersion": "17.0.7",
+  "kindImplementer": "Red Hat, Inc.",
+  "name": "Quarkus",
+  "version": "2.13.8.Final-redhat-00004"
+}
+```
+
+### Workflow Description
+
+In its existing implementation, the OpenShift Insights Operator periodically gathers workload data every 12 hours.
+
+The Operator iterates over all containers in the cluster and captures their Image ID and entry points.
+At that point, the operator would use a “container scanner” to extract information from running containers and augment the gathered data for its workload.
+
+![container scannner workflow](./container-scanner-diagram-1.png "Container Scanner Worklow")
+
+This is not a user-facing feature. It is achieved within the OpenShift Insights Operator gathering workflow.
+The user is able to examine the gathered data by unarchiving the archive generated by the operator.
+The DaemonSet for the container scanner is deployed on-demand when the operator gathers workload data and is not accessible outside of the openshift-insights namespace.
+
+From the operator perspective, the Container Scanner is a black box, and its requirements are:
+
+* Deploy the container scanner `DaemonSet` when gathering workload is starting
+  * The DaemonSet has readiness probes to determine when it is ready to be used. The operator must wait until it is ready and if that takes too long, gather workload without scanning the containers. In that case, an error will be reported by the operator in the archive’s gathers.json file in the `"workloads/workload_info"` section.
+  * The container scanner image must be present in the cluster at installation. The existing tolerations of the Insights Operator should be sufficient to cover the container scanner execution..
+* Determine the Pod IP address of each container-scanner’s Pod (as the container scanner is deployed as a DaemonSet, there is a single pod instance on each worker node) and map it to its worker node.
+  * There is no service exposing the Container Scanner and its pods are accessed directly from their Pod IP address 
```

Comment:
I think in terms of performance it could be problematic. In some larger clusters it would mean connecting dozens or hundreds of Pods.

Reply:
I'm not sure I follow you, we would have as many container scanner pods as there are worker nodes in the cluster and we need to access them individually to scan containers running on the same worker node.

Reply:
yes that's what I am saying. I think this will be quite time consuming in bigger clusters.

Reply:
Can we aggregate and gather data for all pods on a node in a single pass instead?

Reply:
that's definitely achievable, the container scanner could gather all data in a `namespace`/`pod`/`container` tree when workload gather starts and then populates the payload as it goes.
This would remove the multiple HTTP requests, and can be parallelized inside the container scanner code which would result in a much faster time.

Reply:
I've updated the RFE to gather all containers runtime info in a single pass when the workload gatherer starts.